packages/endo/src/node-module-specifier.js

```diff
@@ -0,0 +1,114 @@
+// q, as in quote, for error messages.
+const q = JSON.stringify;
+
+const isRelative = spec =>
+  spec.startsWith("./") ||
+  spec.startsWith("../") ||
+  spec === "." ||
+  spec === "..";
+
+const normalize = (parts, path) => {
+  for (const part of path) {
+    if (part === "." || part === "") {
+      // no-op
+    } else if (part === "..") {
+      if (parts.length === 0) {
+        return undefined;
+      }
+      parts.pop();
```

Comment:
Mutating the caller's argument in-place is a bit surprising, especially because you're also returning it.